src/Microsoft.AspNetCore.Routing/EndpointFeature.cs

```diff
@@ -6,12 +6,39 @@
 
 namespace Microsoft.AspNetCore.Routing
 {
-    public sealed class EndpointFeature : IEndpointFeature
+    public sealed class EndpointFeature : IEndpointFeature, IRoutingFeature
     {
+        private RouteData _routeData;
+        private RouteValueDictionary _values;
+
         public Endpoint Endpoint { get; set; }
 
         public Func<RequestDelegate, RequestDelegate> Invoker { get; set; }
 
-        public RouteValueDictionary Values { get; set; }
+        public RouteValueDictionary Values
+        {
+            get => _values;
+            set
+            {
+                _values = value;
+
+                // RouteData will be created next get with new Values
+                _routeData = null;
```

Comment:
Best attempt to keep the two RVD in sync I could think of. Do they change during a request much?

Reply:
Yeah, this seems fine now that it's not making a copy. It's not common to **set** the values, but it's common to modify the dictionary itself.